src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -1354,13 +1361,31 @@ public sealed override async Task<HttpResponseMessage> SendAsync(HttpRequestMess
             try
             {
                 // Send headers
-                http2Stream = await SendHeadersAsync(request, cancellationToken).ConfigureAwait(false);
-
-                // Send request body, if any
-                await http2Stream.SendRequestBodyAsync(cancellationToken).ConfigureAwait(false);
+                bool shouldExpectContinue = request.Content != null && request.HasHeaders && request.Headers.ExpectContinue == true;
 
-                // Wait for response headers to be read.
-                await http2Stream.ReadResponseHeadersAsync().ConfigureAwait(false);
+                http2Stream = await SendHeadersAsync(request, cancellationToken, flush: shouldExpectContinue).ConfigureAwait(false);
+                if (shouldExpectContinue)
+                {
+                    // Send header and wait a little bit to see if server sends 100, reject code or nothing.
+                    if (NetEventSource.IsEnabled) Trace($"Request content is not null, start processing 100-Continue.");
+                    await http2Stream.SendRequestBodyWithExpect100ContinueAsync(cancellationToken).ConfigureAwait(false);
+                }
+                else
+                {
+                    // Send request body, if any
+                    Task bodyTask = http2Stream.SendRequestBodyAsync(cancellationToken);
+                    if (!request.SendConcurent)
+                    {
+                        await bodyTask.ConfigureAwait(false);
+                        bodyTask = null;
+                    }
+                    // Wait for response headers to be read.
+                    await http2Stream.ReadResponseHeadersAsync().ConfigureAwait(false);
+                    if (bodyTask != null && bodyTask.IsCompleted && bodyTask.IsFaulted)
+                    {
+                        throw bodyTask.Exception;
```

Comment:
This should be:
```C#
await bodyTask.ConfigureAwait(false);
```

Reply:
Also, similar to above, we should null out the bodyTask before we await it, so that elsewhere we can avoid processing the bodyTask if it's already bee handled, e.g.
```C#
if (bodyTask != null && bodyTask.IsCompleted)
{
    Task t = bodyTask;
    bodyTask = null;
    await t.ConfigureAwait(false);
}
```